src/Framework/PaymentGateways/PaymentGateway.php

```diff
@@ -57,6 +56,42 @@ public function __construct(SubscriptionModule $subscriptionModule = null)
         $this->subscriptionModule = $subscriptionModule;
     }
 
+    /**
+     * @unreleased
+     */
+    public static function supportsApiVersions(): array
+    {
+        $versions = [];
+
+        if (method_exists(static::class, 'getLegacyFormFieldMarkup')) {
+            $versions[] = 2;
+        }
+
+        return $versions;
+    }
```

Comment:
Could this simply be a static `$supportedApiVersions = [2]` property? Ideally, it would be great if this method simply knew how to identify a v3 gateway. I'm still considering a `PaymentGatewayV3` interface — even if its empty — that informs this function that it supports v3.

Otherwise, a property seems sufficient.

Reply:
The only problem with properties is we can't enforce them in the `PaymentGatewayInterface`

Reply:
So, i've been thinking about this and I think we need to have a further discussion.  Api versions will not always be additions, but more often - changes to the existing api.  A great example of this would be

**V2**
```
public function createPayment(Donation $donation, $gatewayData);
```

**V3**
```
public function createPayment(Donation $donation, array $gatewayData);
```

This brings up a few questions.
1a. Should each version interface specify _all_ the necessary gateway methods?
1b. If so, wouldn't gateways need to both extend the main gateway class and implement an api version?
1c. Along with this, the existing `PaymentGatewayInterface` would need to be v2 but is currently being implemented by the main gateway class which would cause conflicts for additional version interfaces being implemented.  Also, we have to think about the `SubscriptionModuleInterface` that this one extends.
2. Should we just create a new V3 namespace for a new base gateway class that implements the v3 interfaces?

We want to avoid breaking changes to existing gateways but also be able to enforce new api updates when opted in.  From the questionnaire above we can see where things get a little dicey with using interfaces to specify versions. Personally, I'm leaning toward `#2` to simplify things for gateways but that comes with it's own caveats as well.

Reply:
I was actually having the exact same thought upon reflection yesterday. If we're going to version, it makes sense to me to have an interface that reflects all the necessary methods for that version. A gateway _can_ implement both **if that is possible**. We also have things like the `static::id()` and `$this->getId()` that can be cleaned up (getting rid of the latter).

I'm cool with both having a v2/v3 interface _and_ a v3 namespace with its own base class that implements the v3 interface. The base class ultimately isn't meant to be necessary (implementing code should reference the interface, not the base class), but a super easy starting point that also sets up the subscription module system for us (that really isn't necessary).